### PR TITLE
teach spack how to build pmix from git

### PR DESCRIPTION
**Pull request by roblatham00**

### CODE REVIEW

**Comment by adamjstewart** [comment]
@rhc54 can you review this PR?

File: var/spack/repos/builtin/packages/pmix/package.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 from spack import *
 import spack.architecture
+import os
+import sys
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pmix/package.py]: Flake8:
```
var/spack/repos/builtin/packages/pmix/package.py:10: [F401] 'sys' imported but unused
```

File: var/spack/repos/builtin/packages/pmix/package.py [hunk 2]
```diff
@@ -49,6 +53,20 @@ class Pmix(AutotoolsPackage):
...
     depends_on('libevent@2.0.20:2.0.22,2.1.8')
     depends_on('hwloc@1.11.0:1.11.99,2.0.1:', when='@3.0.0:')
+    depends_on("m4", type=("build"), when="@master")
+    depends_on("autoconf", type=("build"), when="@master")
+    depends_on("automake", type=("build"), when="@master")
+    depends_on("libtool", type=("build"), when="@master")
+    depends_on("perl", type=("build"), when="@master")
+
+    def autoreconf(self, spec, prefix):
+        """Only neeed when building from git checkout"""
```
Comment (rhc54) [inline var/spack/repos/builtin/packages/pmix/package.py]: Minor spelling nit: "neeed" -> "need"

**Review by rhc54 — COMMENTED** [review]
PMIx master has a couple of optional dependencies on libcurl and libjansson (version 2.11 or greater). I don't know if you need to reflect that in spack or not as those are only active for optional features if the libraries are available and the user requests those features, but just wanted to alert you to it.

**Comment by roblatham00** [comment]
@rhc54 i have added a variant for the curl and jansson features.  I've called it '+restful' but we can call it anything that makes sense to you. 

I made the 'restful' variant conflict with anything 3.1.5 or newer: is that in line with your release roadmap ?

**Comment by rhc54** [comment]
"restful" sounds fine. The restful support will start with v4.0 - it isn't going to be backported to any earlier release series. I'm thinking v4.0 will be released in July.

**Comment by roblatham00** [comment]
ok latest revision conflicts 'restful' with anything in the 3 series

**Review by rhc54 — APPROVED** [review]
Thanks - looks good to me!

File: var/spack/repos/builtin/packages/pmix/package.py [hunk 3]
```diff
@@ -49,11 +52,34 @@ class Pmix(AutotoolsPackage):
...
+        perl('./autogen.pl')
 
     variant('pmi_backwards_compatibility',
             default=True,
             description="Toggle pmi backwards compatibility")
 
+    variant('restful',
+            default=False,
+            description="allow a PMIx server to request services from "
+            "a system-level REST server")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pmix/package.py]: Is there a corresponding configure flag to enable/disable this feature?

File: var/spack/repos/builtin/packages/pmix/package.py [hunk 4]
```diff
@@ -49,11 +52,34 @@ class Pmix(AutotoolsPackage):
...
+    depends_on('curl', when="+restful")
+    depends_on('jansson@2.11:', when="+restful")
+
+    conflicts('@:3.9.9', when='+restful')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pmix/package.py]: This isn't required, but I usually like to sort things in packages like:

1. Docstring
2. URLs
3. Versions
4. Variants
5. Dependencies
6. Conflicts
7. Functions

I would definitely move autoreconf down above configure_args, and probably move the variant between versions and dependencies.

File: var/spack/repos/builtin/packages/pmix/package.py (hunk 3 above)
Comment (rhc54) [reply var/spack/repos/builtin/packages/pmix/package.py]: `--with-curl --with-libjansson` plus the usual `=<path>` for non-standard locations. Both options will automatically enable if the support is found in the standard locations, so I will likely add something like `--enable-restful-support` to make it easier to turn "off" if someone doesn't want it.

File: var/spack/repos/builtin/packages/pmix/package.py (hunk 4 above)
Comment (roblatham00) [reply var/spack/repos/builtin/packages/pmix/package.py]: thanks for the guidance.  reordered things in a third commit

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**